Loop/Managers/LoopDataManager.swift

```diff
@@ -1078,6 +1092,98 @@ extension LoopDataManager {
         return prediction
     }
 
+    fileprivate func predictGlucoseFromManualGlucose(
+        _ glucose: NewGlucoseSample,
+        potentialBolus: DoseEntry?,
+        potentialCarbEntry: NewCarbEntry?,
+        replacingCarbEntry replacedCarbEntry: StoredCarbEntry?,
+        includingPendingInsulin: Bool
+    ) throws -> [PredictedGlucoseValue] {
+        let retrospectiveStart = glucose.date.addingTimeInterval(-retrospectiveCorrection.retrospectionInterval)
+        let earliestEffectDate = Date(timeIntervalSinceNow: .hours(-24))
+        let nextEffectDate = insulinCounteractionEffects.last?.endDate ?? earliestEffectDate
+
+        let updateGroup = DispatchGroup()
+        let effectCalculationError = Locked<Error?>(nil)
+
+        var insulinEffect: [GlucoseEffect]?
+        let basalDosingEnd = includingPendingInsulin ? nil : Date()
+        updateGroup.enter()
+        doseStore.getGlucoseEffects(start: nextEffectDate, basalDosingEnd: basalDosingEnd) { result in
+            switch result {
+            case .failure(let error):
+                effectCalculationError.mutate { $0 = error }
+            case .success(let effects):
+                insulinEffect = effects
+            }
+
+            updateGroup.leave()
+        }
+
+        updateGroup.wait()
+
+        var insulinCounteractionEffects = self.insulinCounteractionEffects
+        if nextEffectDate < glucose.date, let insulinEffect = insulinEffect {
+            updateGroup.enter()
+            glucoseStore.getCounteractionEffects(start: nextEffectDate, to: insulinEffect, unstoredGlucoseSample: glucose) { velocities in
+                insulinCounteractionEffects.append(contentsOf: velocities)
+                insulinCounteractionEffects = insulinCounteractionEffects.filterDateRange(earliestEffectDate, nil)
+
+                updateGroup.leave()
+            }
+
+            updateGroup.wait()
+        }
+
+        var carbEffect: [GlucoseEffect]?
+        updateGroup.enter()
+        carbStore.getGlucoseEffects(
+            start: retrospectiveStart,
+            effectVelocities: settings.dynamicCarbAbsorptionEnabled ? insulinCounteractionEffects : nil
+        ) { result in
+            switch result {
+            case .failure(let error):
+                effectCalculationError.mutate { $0 = error }
+            case .success(let (_, effects)):
+                carbEffect = effects
+            }
+
+            updateGroup.leave()
+        }
+
+        updateGroup.wait()
+
+        if let error = effectCalculationError.value {
+            throw error
+        }
+
+        return try predictGlucose(
+            startingAt: glucose.quantitySample,
+            using: [.insulin, .carbs],
```

Comment:
Momentum isn't relevant since CGM data is absent; RC _could_ be relevant if glucose data is < 30min old, though last I spoke with Product the ask was to use only insulin and carbs for this prediction.